bundle/store.go

```diff
@@ -322,24 +336,90 @@ func activateBundles(opts *ActivateOpts) error {
 		remainingAndExtra[name] = mod
 	}
 
-	err = writeModules(opts.Ctx, opts.Store, opts.Txn, opts.Compiler, opts.Metrics, opts.Bundles, remainingAndExtra, opts.legacy)
+	err = writeModules(opts.Ctx, opts.Store, opts.Txn, opts.Compiler, opts.Metrics, snapshotBundles, remainingAndExtra, opts.legacy)
 	if err != nil {
 		return err
 	}
 
-	for name, b := range opts.Bundles {
-		// Always write manifests to the named location. If the plugin is in the older style config
-		// then also write to the old legacy unnamed location.
-		if err := WriteManifestToStore(opts.Ctx, opts.Store, opts.Txn, name, b.Manifest); err != nil {
+	for name, b := range snapshotBundles {
+		if err := writeManifestToStore(opts, name, b.Manifest); err != nil {
+			return err
+		}
+
+		if err := writeWasmModulesToStore(opts.Ctx, opts.Store, opts.Txn, name, b); err != nil {
 			return err
 		}
-		if opts.legacy {
-			if err := LegacyWriteManifestToStore(opts.Ctx, opts.Store, opts.Txn, b.Manifest); err != nil {
+	}
+
+	return nil
+}
+
+func activateDeltaBundles(opts *ActivateOpts, bundles map[string]*Bundle) error {
+
+	// Check that the manifest roots and wasm resolvers in the delta bundle
+	// match with those currently in the store
+	for name, b := range bundles {
+		value, err := opts.Store.Read(opts.Ctx, opts.Txn, ManifestStoragePath(name))
+		if err != nil {
+			return err
+		}
+
+		manifest, ok := value.(Manifest)
+		if !ok {
+			return fmt.Errorf("corrupt manifest")
+		}
+
+		if len(b.Manifest.WasmResolvers) != len(manifest.WasmResolvers) {
+			return fmt.Errorf("delta bundle %s has manifest wasm resolvers that are different from those in the store", name)
+		}
+
+		for i := 0; i < len(b.Manifest.WasmResolvers); i++ {
+			if b.Manifest.WasmResolvers[i] != manifest.WasmResolvers[i] {
+				return fmt.Errorf("delta bundle %s has manifest wasm resolvers that are different from those in the store", name)
+			}
+		}
+
+		if !b.Manifest.rootSet().Equal(manifest.rootSet()) {
+			return fmt.Errorf("delta bundle %s has manifest roots that are different from those in the store", name)
+		}
+	}
+
+	for _, b := range bundles {
+		for _, patch := range b.Patch.Data {
```

Comment:
I think this could be factored into another function that accepts a series of patches and a store/txn as input and applies them.

Reply:
Done